### DISCUSSION TITLE
Buffered datagram receipt

### DISCUSSION THREAD

**Discussion by wegylexy** [microsoft/msquic#1654] [Q&A] [answered]
Is it possible to keep the received datagram buffer alive? This will enable something like `ReceiveDatagramAsync()` in .NET 6 without having to copy the buffer content, and align with the async stream API.
For streams, it is possible to return `QUIC_STATUS_PENDING` in the handler to hold the buffer.
If the datagram buffer is on the stack though, I would propose dispatching an event with `ReadOnlySpan<byte>` in .NET 6 to avoid heap allocation.

**Comment by nibanks** [comment, marked as answer]
This would be tricky to pull off. Currently, datagram buffers (pointers) are directly from the UDP datagrams. We really don't want to hold on to those for any length of time. If we wanted to support an async model, then we'd have to pend the UDP datagrams while we wait for the app to complete the receives. It's doable, but I would first prefer to measure the impact of not doing it (and just copying the data as necessary).
